internal/services/payment_to_submitter_service.go

```diff
@@ -157,7 +152,57 @@ func (s PaymentToSubmitterService) sendPaymentsReadyToPay(
 	return nil
 }
 
-func (s PaymentToSubmitterService) sendPaymentsToTSS(ctx context.Context, tssDBTx db.DBTransaction, tenantID string, pendingPayments []*data.Payment) error {
+func (s PaymentToSubmitterService) sendPaymentsToProperPlatform(ctx context.Context, sdpDBTx, tssDBTx db.DBTransaction, tenantID string, paymentsToSubmit []*data.Payment) error {
+	distAccount, err := s.distAccountResolver.DistributionAccountFromContext(ctx)
+	if err != nil {
+		return fmt.Errorf("getting distribution account: %w", err)
+	}
+
+	if distAccount.IsCircle() {
+		return s.sendPaymentsToCircle(ctx, sdpDBTx, distAccount.CircleWalletID, paymentsToSubmit)
+	} else {
+		return s.sendPaymentsToTSS(ctx, sdpDBTx, tssDBTx, tenantID, paymentsToSubmit)
+	}
+}
+
+func (s PaymentToSubmitterService) sendPaymentsToCircle(ctx context.Context, sdpDBTx db.DBTransaction, circleWalletID string, paymentsToSubmit []*data.Payment) error {
+	for _, payment := range paymentsToSubmit {
+
+		// 1. Persist the circle idempotency key if needed
+		idempotencyKey, err := s.sdpModels.Payment.SetCircleIdempotencyKeyIfNeeded(ctx, sdpDBTx, payment.ID)
+		if err != nil {
+			return fmt.Errorf("setting circle idempotency key: %w", err)
+		}
+
+		// 2. Submit the payment to Circle
+		transfer, err := s.circleService.SendPayment(ctx, circle.PaymentRequest{
+			SourceWalletID:            circleWalletID,
+			DestinationStellarAddress: payment.ReceiverWallet.StellarAddress,
+			Amount:                    payment.Amount,
+			StellarAssetCode:          payment.Asset.Code,
+			IdempotencyKey:            idempotencyKey,
+		})
+
+		if err != nil {
+			// 3. If the transfer fails, set the payment status to failed
+			// TODO:  If the transfer fails because of authentication error, set the account status to `PENDING_USER_ACTIVATION` - [SDP-1245]
+			paymentErr := s.sdpModels.Payment.MarkAsFailed(ctx, sdpDBTx, payment.ID, err.Error())
+			if paymentErr != nil {
+				return fmt.Errorf("marking payment as failed: %w", paymentErr)
+			}
+		} else {
+			// 4. If sending transfer request succeeds, update the payment status to pending and store the transfer id
+			err = s.sdpModels.Payment.MarkCirclePaymentAsPending(ctx, sdpDBTx, payment.ID, transfer.ID)
+			if err != nil {
+				return fmt.Errorf("marking payment as pending: %w", err)
+			}
+			log.Ctx(ctx).Infof("Submitted payment %s to Circle: %+v", payment.ID, transfer)
+		}
+	}
+	return nil
+}
```

Comment:
👋  To avoid double-spending, I believe the `SetCircleIdempotencyKeyIfNeeded` should use the dbConnectionPool rather than the database transaction.

Reply:
Also, the most important function to add thorough tests is this one. It's important to understand the behavior of the system when some payments succeed and others fail.

Reply:
addressed in d359772de9ad6379f88d9f04e3065b7d419c75fd